Use a Group to identify mCODE-conforming patients

masnick opened the pull request:
This PR profiles Group to identify mCODE-conforming Patient resources in the case where the implementer does not want all patients with cancer Conditions to conform to mCODE.

I also made two structural changes to facilitate this:

1. The CapabilityStatement-related FSH is now in 3 separate files:
    1. `DEF_Capabilities.fsh` has the the Profile and RuleSet used by the other files
    2. `DEF_Capabilities_Server.fsh` has all the Server-related FSH
    3. `DEF_Capabilities_Client.fsh` is now generated from `DEF_Capabilities_Server.fsh` with a series of find/replace operations
2. The UML diagrams in `conformance.md` are now automatically regenerated and inserted inline by publisher.jar.

The conformance documentation is updated to reflect the change from `meta.profile` to using Group.

masnick commented on input/fsh/DEF_Capabilities.fsh:
```diff
@@ -36,267 +38,4 @@ RuleSet: mCODECapabilityStatementRules
-Instance: mcode-receiver-fallback2
-InstanceOf: MCodeCapabilityStatement
-Usage: #definition
-Description: "Defines the secondary fallback requirements for an mCODE Data Receiver"
-* name = "MCODEReceiverFallback2"
-* title = "mCODE Data Receiver"
-* description = "Defines the secondary fallback requirements for the mCODE Data Receiver"
-* insert mCODECapabilityStatementRules
-
-* insert mCODE_CS_Client
-
-// GET [base]/Condition?code:in=http://hl7.org/fhir/us/mcode/ValueSet/mcode-primary-or-uncertain-behavior-cancer-disorder-vs&_include=Condition:subject
-* insert mCODE_CS_Client_Fallback
-
-* rest[clientSlice].resource[ConditionSlice].searchInclude[0] = "Condition:subject"
-
-///////////////////////////////////////////////////////////////////////////////////////////////////
-
-Instance: mcode-receiver-fallback3
-InstanceOf: MCodeCapabilityStatement
-Usage: #definition
-Description: "Defines the tertiary fallback requirements for an mCODE Data Receiver"
-* name = "MCODEReceiverFallback3"
-* title = "mCODE Data Receiver"
-* description = "Defines the tertiary fallback requirements for the mCODE Data Receiver"
-* insert mCODECapabilityStatementRules
-
-* insert mCODE_CS_Client
-
-// GET [base]/Condition?code:in=http://hl7.org/fhir/us/mcode/ValueSet/mcode-primary-or-uncertain-behavior-cancer-disorder-vs
-* insert mCODE_CS_Client_Fallback
-
-// GET [base]/Patient?_id=id1|id2|id3
-* rest[clientSlice].resource[PatientSlice].searchParam[0].name = "_id"
-* rest[clientSlice].resource[PatientSlice].searchParam[0].type = #token
-* rest[clientSlice].resource[PatientSlice].searchParam[0].definition = "http://hl7.org/fhir/us/mcode/SearchParameter/Patient-id"
-* rest[clientSlice].resource[PatientSlice].searchParam[0].extension.url = "http://hl7.org/fhir/StructureDefinition/capabilitystatement-expectation"
-* rest[clientSlice].resource[PatientSlice].searchParam[0].extension.valueCode = #SHALL
-
-///////////////////////////////////////////////////////////////////////////////////////////////////
```
This is split into two separate files now for the Client and Server.

masnick commented on input/fsh/SD_Group.fsh:
```diff
@@ -0,0 +1,31 @@
+// "Identifies traits whose presence or absence is shared by members of the group."
+// https://www.hl7.org/fhir/group-definitions.html#Group.characteristic.code
+* characteristic 1..*
+* characteristic.code = #administrative
+* characteristic.valueCodeableConcept = #mcode-cancer-patient "mCODE Cancer Patient"
```
Specifying `characteristic.valueCodeableConcept` is the only way I could see to cleanly identify a `Group` resource as conforming to this profile.

I don't have a strong opinion about the `valueCodeableConcept` value that's used here. If there's a better option than what I have, please let me know.

masnick commented on input/images/mcode-patient-bundle-pull.svg:
```diff
@@ -1,20 +0,0 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="159px" preserveAspectRatio="none" style="width:526px;height:159px;" version="1.1" viewBox="0 0 526 159" width="526px" zoomAndPan="magnify"><defs><filter height="300%" id="fumpyee5pov0 ...
-@startuml
-skinparam defaultFontName Arial
-
-title mCODE Patient Bundle: Pull Model
-hide footbox
-participant "mCODE Data Sender" as Sender
-participant "mCODE Data Receiver" as Receiver
-"Receiver" -> "Sender" : ""GET [base]/Patient/[id]/$mcode-everything""
-"Sender" - -> "Receiver" : HTTP response containing mCODE Patient Bundle
```
Removed these `.svg` files because publisher.jar is now inserting the SVG inline into the HTML. It does this based on a naming convention: any files in `input/images-source` with a `.plantuml` extension get picked up by the publisher and can be inserted in a `.md` file with `<div>{%include some_filename.svg%}</div>`.

markkramerus commented:
The indentation in conformance.html is really out of control.  In Section 2.2.2.1 (four layers deep), the steps are numbered an additional three levels deep. You have to figure out a simpler arrangement of information, and use a system of step numbering where you don't have levels that are singly-numbered.  Either it has to be numbered to explicitly show depth (e.g. 1.1.1) or numbered with a numeric level followed by an alphabetic level followed by a roman numeral (1.  then a.  then i.)

input/pagecontent/conformance.md

markkramerus commented:
2) "Retrieve mCODE Patient. mCODE Data Senders SHALL implement the following operation for retrieving a Patient resource by id UNLESS they are a specialty system that does not implement CancerPatient due to unavailable data as described above," 

Is it possible to NOT implement CancerPatient and still be mCODE conformant? Would that mean there would be no "mCODE Patients"?  If there are no mCODE patients, there is nothing left to conforming to any conformance rules. Am I misreading this?

markkramerus commented:
3. In the sequence diagrams, you have the sender on the left. Usually, by convention, the client is on the left, since it initiates the sequence. It is a pretty strong convention in my experience. Am I wrong about this?

markkramerus commented:
4. Still on conformance page, you have to say if the Group method of finding mCODE patients is implemented, it takes precedence. You might consider it the primary way to get mCODE patients, with searches on patients with cancer the fallback. I'm not really saying one is preferred over another, but in practice, a client would always have to test if there is an mcode-patient-group (however, if Group resource is not supported, that's a different story).

markkramerus commented:
5. Where you are describing using Group, there is some html breaking through:
```
<div style="text-align: center;"><?xml version=”1.0” encoding=”UTF-8” standalone=”no”?>
```

markkramerus commented:
6. In 2.2.2.2, you say "mCODE Data Receivers SHALL be able to initiate ALL the requests described above. Additionally, they SHALL be able to read and process ALL individual resources returned by the operations above that conform to mCODE profiles identified as supported by their CapabilityStatements.".  
Is "their" referring to the receiver or sender's CapabilityStatement?  Please clarify.

markkramerus commented:
7. In 2.2.2.3, I guess I'm not clear on what it means to comply with a CapabilityStatement. Is that defined somewhere? Is there a validator one can use to check conformance? (Since the mcode-sender-preferred and its kin are instances, it isn't like testing an instance against a profile.)

markkramerus commented:
8. Section 2.3 provides the motivation for much of the content in 2.2.2.x. I'm wondering if you can re-organize the information so the content of 2.3 is discussed together with the ways to implement it.

markkramerus commented:
9. In 2.9, you say "mCODE’s rules regarding required data elements are **the same as** US Core’s rules. To paraphrase, a Data Sender must provide each required element or an explicit data absent reason for each missing data item, and cannot substitute a nonsense or filler value just to satisfy the data requirement."
Are the cardinality rules given just above this statement **the same as** US Core's rules? You might clarify that the cardinality rules are FHIR's rules, not US Core. Reading FHIR's documentation (https://www.hl7.org/fhir/conformance-rules.html#cardinality),  FHIR never really comes out and says what you say here. I guess they are just relying on "common sense".

markkramerus commented:
10. It is pretty hard to find the capability statements. They are buried pretty deep on the Artifacts page in the middle of a bunch of examples. Can you adjust `sushi-config.yaml` to bring them closer to the top of the page?

markkramerus commented:
11.  Why do you need the profile "mCODE Capability Statement"?  Is it just to help you make rule sets that don't have fixed indices? Are there any conformance expectations associated with this profile?  I think it is confusing to have the profile unless it is somehow part of the conformance criteria.

markkramerus commented:
12. I don't think you need to profile Group. You might be able to drop this profile outright.
I think all that is needed is to either fix the code (or characteristic), and make the query with that code (or characteristic) part of the conformance requirements. Between code and characteristic, I prefer code. The documentation does not forbid using code (it just says "generally") and usually `code` is what queries are based on. A code like "mcode-patient" would work to identify the Group resource containing mcode patients.

markkramerus commented:
13. In 8.0.14, the descriptions should be improved. You should differentiate profiles from instances, and state what something is (profile or instance), what is represents (e.g., a requirement capability statement), and what it is for. 
Also, this doesn't really parse: "Primary fallback capabilities for an mCODE Data Receiver where all cancer patients conform to mCODE."

markkramerus commented:
14. I would like to reconsider the names of the capability statements. They should be named by function than by priority. Might these names be clearer?
mcode-sender-subset --> mcode-sender-patients-in-group
mcode-sender-preferred -->  mcode-sender-patients-with-cancer-condition
mcode-sender-fallback1 --> mcode-sender-patients-and-cancer-conditions
mcode-sender-fallback2 --> mcode-sender-cancer-conditions-then-patients

markkramerus commented:
15. Are the capability statements complete? Certainly there would be other resources, search parameters, and profiles. You haven't made it clear that these represent a minimal "find mCODE patients" portion.
Should we include a "maximal CP" that has all the mCODE profiles, and all the mCODE (or US Core) search parameters?

markkramerus reviewed (changes requested):
See separate comments.

masnick commented:
> 1) The indentation in conformance.html is really out of control. In Section 2.2.2.1 (four layers deep), the steps are numbered an additional three levels deep. You have to figure out a simpler arrangement of information, and use a system of step numbering where you don't have levels that are singly-numbered. Either it has to be numbered to explicitly show depth (e.g. 1.1.1) or numbered with a numeric level followed by an alphabetic level followed by a roman numeral (1. then a. then i.)
>
> 8) Section 2.3 provides the motivation for much of the content in 2.2.2.x. I'm wondering if you can re-organize the information so the content of 2.3 is discussed together with the ways to implement it.

👍 I've re-organized/re-written the entire `conformance.html` page, simplifying its structure substantially. There are now 4 main sections:

1. Roles and architecture
2. MUST-level conformance criteria
3. SHOULD-level conformance criteria
4. General profile-level conformance criteria

There deepest heading level is now `<h4>`, which is functionally only **one** level of sub-headings within the 4 sections listed above.

> 2) "Retrieve mCODE Patient. mCODE Data Senders SHALL implement the following operation for retrieving a Patient resource by id UNLESS they are a specialty system that does not implement CancerPatient due to unavailable data as described above,"
> 
> Is it possible to NOT implement CancerPatient and still be mCODE conformant? Would that mean there would be no "mCODE Patients"? If there are no mCODE patients, there is nothing left to conforming to any conformance rules. Am I misreading this?

My thinking here was that some participants, like genomics labs, might only want to use the genomics-related mCODE profiles. However, I'm personally fine with these participants not conforming to mCODE -- this of course doesn't stop them from implementing whatever parts of mCODE they want to. Additionally, the "escape hatch" language like what you quote above just adds confusion.

I've therefore required support for CancerPatient and PrimaryCancerCondition as MUST, as well as the CapabilityStatements for identifying CancerPatient-conforming resources and retrieving a single CancerPatient.

> 3) In the sequence diagrams, you have the sender on the left. Usually, by convention, the client is on the left, since it initiates the sequence. It is a pretty strong convention in my experience. Am I wrong about this?

Thanks for pointing this out; fixed.

> 4) Still on conformance page, I think you need to say - if the Group method of finding mCODE patients is implemented, it takes precedence. You might posit it as the primary way to get mCODE patients, whereas, searches on patients with cancer the fallback. I'm not really saying one is preferred over another, but in practice, wouldn't a client have to test if there is an mcode-patient-group (or would this be known from the CapabilityStatement?).

I've addressed this by requiring all participants to support `GET [base]/Group?code=mcode-patient` with either 0 or 1 result, and making this the preferred way to identify mCODE Patients. The non-Group operations are now fallbacks, and this is all in a single (one-level) list explicitly ordered by preference.

> 5) Where you are describing using Group, there is some html breaking through.

Whoops, fixed.

> 6) In 2.2.2.2, you say "mCODE Data Receivers SHALL be able to initiate ALL the requests described above. Additionally, they SHALL be able to read and process ALL individual resources returned by the operations above that conform to mCODE profiles identified as supported by their CapabilityStatements.".
> Is "their" referring to the receiver or sender's CapabilityStatement? Please clarify.

Fixed.

> 7) In 2.2.2.3, I guess I'm not clear on what it means to comply with a CapabilityStatement. Is compliance to a CS defined somewhere? Is there a validator that can check compliance? (Since the mcode-sender-preferred and its kin are instances, it isn't like testing an instance against a profile.)

This section was removed in the rewrite.

> 9) In 2.9, you say "mCODE’s rules regarding required data elements are the same as US Core’s rules. To paraphrase, a Data Sender must provide each required element or an explicit data absent reason for each missing data item, and cannot substitute a nonsense or filler value just to satisfy the data requirement."
> Are the cardinality rules given just above this statement the same as US Core's rules? You might clarify that the cardinality rules are FHIR's rules, not US Core. Reading FHIR's documentation (https://www.hl7.org/fhir/conformance-rules.html#cardinality), FHIR never really comes out and says what you say here. I guess they are just relying on "common sense".

I have no memory of writing this, but I looked at the US Core link from conformance.md and the FHIR link you included above. I agree that what we're saying here does not directly come from either, but I could _swear_ I read something about not using nonsense/filler data somewhere. I can't find it now though. I've tweaked the language to avoid confusion.

> 10) It is pretty hard to find the capability statements. They are buried pretty deep on the Artifacts page in the middle of a bunch of examples. Can you adjust sushi-config.yaml to bring them closer to the top of the page?

For reference, the current ordering:

1. Profiles
2. Abstract profiles
3. Extensions
4. Value sets & code systems
5. Examples
6. Conformance resources
7. Experimental resources

I'm not sure how far up we want to move the conformance resources, or if we want to re-order anything else. I personally don't feel strongly about the ordering here other than that profiles should be at the top.

This does raise a bigger problem: the Artifacts Summary page is a bear to navigate because it's so long. Following [US Core's](https://www.hl7.org/fhir/us/core/) example, I adjusted our nav bar to have a "FHIR Artifacts" dropdown that either links to the Artifacts Summary page or to the existing group-specific pages (examples, experimental resources). 

I don't think this is the ideal solution. It would be better if the template were improved, and there's some discussion [on Zulip](https://chat.fhir.org/#narrow/stream/179252-IG-creation/topic/Base.20template.20suggestion(s)) about this. But for now I think the dropdown helps.

What do you think about this change?

![image](https://user-images.githubusercontent.com/110426/103765260-e356f300-4fea-11eb-828b-ea98c05edec7.png)

> 11) Why do you need the profile "mCODE Capability Statement"? Is it just to help you make rule sets that don't have fixed indices? Are there any conformance expectations associated with this profile? I think it is confusing to have the profile unless it is somehow part of the conformance criteria.

👍 Profile removed. Having it made the FSH array indices a bit easier to handle, but I agree the overhead wasn't worth it.

> 12) I don't think you need to profile Group. You might be able to drop this profile outright.
> I think all that is needed is to either fix the code (or characteristic), and make the query with that code (or characteristic) part of the conformance requirements. Between code and characteristic, I prefer code. The documentation does not forbid using code (it just says "generally") and usually code is what queries are based on. A code like "mcode-patient" would work to identify the Group resource containing mcode patients.

👍 Profile removed, switched to `code` rather than `characteristic`.

> 13) In 8.0.14, the descriptions should be improved. You should differentiate profiles from instances, and state what something is (profile or instance), what is represents (e.g., a requirement capability statement), and what it is for.
> Also, this doesn't really parse: "Primary fallback capabilities for an mCODE Data Receiver where all cancer patients conform to mCODE."
> Also, this is not a great name (title?): "Search by _id in Patients with multipleOr"

> 14) I would like to reconsider the names of the capability statements. They should be named by function than by priority. Might these names be clearer?
> mcode-sender-subset --> mcode-sender-patients-in-group
> mcode-sender-preferred --> mcode-sender-patients-with-cancer-condition
> mcode-sender-fallback1 --> mcode-sender-patients-and-cancer-conditions
> mcode-sender-fallback2 --> mcode-sender-cancer-conditions-then-patients

👍 CS are renamed.

> 15) Are the capability statements complete? Certainly there would be other resources, search parameters, and profiles. You haven't made it clear that these represent a minimal "find mCODE patients" portion.
Should we include a "maximal CP" that has all the mCODE profiles, and all the mCODE (or US Core) search parameters?

Yes, I have added a `TODO` placeholder to `conformance.md` to include the FHIR requests to retrieve resources conforming to each mCODE profile. Presumably these could have accompanying CS and search params. I'll tackle this in a separate PR.

masnick commented:
@markkramerus and @mlterryMitre: This is ready for re-review. There are some `TODO` items left in `conformance.md` which I would like to address in separate pull requests after this is merged in if possible.

markkramerus commented:
Great job. I made a few minor editorial changes in conformance.md, but content-wise, I like it. There are still a few to-do's... let's not forget about them.
